packages/affiliates/libs/uniswap/models.ts

```diff
@@ -0,0 +1,169 @@
+import assert from "assert";
+import { exists, getPositionKey } from "./utils";
+import { ethers } from "ethers";
+
+type Id = string | number;
+type HasId = {
+  id?: Id;
+};
+interface MakeId<T> {
+  (obj: T): string;
+}
+
+// Generic table with common functions, store is a map now, but we could replace this with any kind of
+// data store with some minor modifications, as long as it has, set, get, delete, has and values
+export function Table<T extends HasId>(config: { makeId: MakeId<T>; type: string }, store: Map<Id, T>) {
+  const { makeId, type } = config;
+  async function create(data: T) {
+    const id = makeId(data);
+    assert(!(await has(id)), `${type} exists`);
+    return set({ id, ...data });
+  }
+  async function set(data: T) {
+    assert(data.id, `${type} requires id`);
+    await store.set(data.id, { ...data });
+    return data;
+  }
+  async function get(id: string) {
+    assert(await store.has(id), `${type} does not exist`);
+    return store.get(id) as T;
```

Comment:
yes it should, but typescript doesnt recognize that there is a guard statement with the assert above it, and thinks it can possibly return null, which is an incompatible type with T, so this forces it to be T.  its actually quite annoying